core-bundle/src/Util/LocaleUtil.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Util;
+
+/**
+ * The LocaleUtil helps in handling ICU Locale IDs and IETF Language Tags (BCP 47).
+ * Both are almost identical, Locale ID uses underline (_) and Language Tag uses dash (-).
+ *
+ * For any method, we are safely assuming the input can be any format, therefore we try
+ * to handle any input the same.
+ *
+ * @see https://github.com/contao/core-bundle/issues/233
+ */
+class LocaleUtil
+{
+    public static function canonicalize(string $locale): string
+    {
+        return \Locale::canonicalize($locale);
+    }
+
+    /**
+     * Converts an Locale ID (_) to a Language Tag (-) and strips keywords after the @ sign.
+     *
+     * Language Tag is used in two cases in Contao:
+     *  1. The XML/HTML lang attribute.
+     *  2. The _locale attribute in requests/routes, because it might be used to generate legacy routes.
```

Comment:
Didn’t we want to use Locale IDs for the `_locale` attribute because we do that already?

Reply:
<img width="348" alt="Bildschirmfoto 2020-09-16 um 15 51 43" src="https://user-images.githubusercontent.com/367169/93346697-8a453280-f834-11ea-91df-c19ac6a9bda1.png">

Reply:
Or do you mean it is only used there when needed for generating the URL in legacy routing?

Reply:
we can't and we don't do that, because the `_locale` is used in the URL in legacy routing, if `prepend_locale` is enabled. We do convert the `_locale` value to a Locale ID when applying it https://github.com/contao/contao/blob/master/core-bundle/src/EventListener/LocaleSubscriber.php#L94